### PR TITLE
[Runtime] Dynamically select the is-Swift bit at runtime on Apple platforms.

### PR DESCRIPTION
**Pull request by mikeash**
Recent Swift uses 2 as the is-Swift bit when running on newer versions, and 1 on older versions. Since it's difficult or impossible to know what we'll be running on at build time, make the selection at runtime.

### CODE REVIEW

**Comment by mikeash** [comment]
@slavapestov I think you'll appreciate this.

**Comment by mikeash** [comment]
@swift-ci please test

File: include/swift/ABI/Metadata.h [hunk 1]
```diff
@@ -1192,6 +1192,31 @@ struct TargetClassMetadata : public TargetAnyClassMetadata<Runtime> {
...
+  /// may have a different bit set from the one that this runtime canonically
+  /// considers the "is Swift" bit.
+  void setAsTypeMetadata() {
+    // If the wrong "is Swift" bit is set, set the correct one.
+    //
+    // Note that the only time we should see the "new" bit set while
+    // expecting the "old" one is when running a binary built for a
+    // new OS on an old OS, which is not supported, however we do
+    // have tests that exercise this scenario.
+    auto otherSwiftBit = (3ULL - SWIFT_CLASS_IS_SWIFT_MASK);
```
Comment (openloop) [inline include/swift/ABI/Metadata.h]: Enlighten me. No magic numbers.

File: stdlib/public/runtime/BackDeployment.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+SWIFT_ALLOWED_RUNTIME_GLOBAL_CTOR_BEGIN
+extern "C" unsigned long long
+_swift_classIsSwiftMask = computeIsSwiftMask();
+SWIFT_ALLOWED_RUNTIME_GLOBAL_CTOR_END
+#endif // SWIFT_CLASS_IS_SWIFT_MASK_GLOBAL_VARIABLE
+
+static swift::_SwiftNSOperatingSystemVersion swiftInOSVersion = {
+#if __MAC_OS_X_VERSION_MIN_REQUIRED
+  10, 14, 4
```
Comment (openloop) [inline stdlib/public/runtime/BackDeployment.cpp]: These.

**Comment by mikeash** [comment]
https://github.com/apple/swift-lldb/pull/1342

@swift-ci please test

File: stdlib/public/runtime/BackDeployment.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// WatchOS also pretends to be iOS, so check it first.
+#elif __WATCH_OS_VERSION_MIN_REQUIRED
+   5,  2, 0
+#elif __IPHONE_OS_VERSION_MIN_REQUIRED || __TV_OS_VERSION_MIN_REQUIRED
+  12,  2, 0
+#else
+  9999, 0, 0
+#endif
```
[the remark this answers is not in the document]
Comment (mikeash) [reply stdlib/public/runtime/BackDeployment.cpp]: I think it's roughly the same ugliness either way.

File: CMakeLists.txt [hunk 4]
```diff
@@ -235,8 +235,8 @@ option(SWIFT_RUNTIME_CRASH_REPORTER_CLIENT
     FALSE)
 
 option(SWIFT_DARWIN_ENABLE_STABLE_ABI_BIT
-    "Enable the Swift stable ABI's class marker bit"
-    FALSE)
+    "Enable the Swift stable ABI's class marker bit for new deployment targets"
+    TRUE)
```
[the remark this answers is not in the document]
Comment (mikeash) [reply CMakeLists.txt]: Not much point other than just being incremental. I think I'd like to rip out the option as part of a separate task.

**Comment by mikeash** [comment]
apple/swift-lldb#1342

@swift-ci please test

**Comment by compnerd** [comment]
This broke the Windows build!

**Comment by mikeash** [comment]
@compnerd My apologies yet again! I'll be visiting the office the week of March 15th and would love to make it up to you by purchasing you the beverage of your choice if you're around.

**Pull request merged**